### PR TITLE
Closes #643: FullScreen mode support...

### LINKED ISSUE #643: Engine/Session: Fullscreen mode
There are a lot of things needed for fullscreen mode. For this issue I propose to only focus on the `browser-session` and `browser-engine` components and ignore the actual UI pieces. We will need the `browser-session` implementation for Focus parity (#200). We should do the `browser-engine` work too to come up with the right state representation in `browser-session`.

The APIs of GeckoView and WebView are fundamentally different which will make this more difficult to implement:
* WebView will return a `View` object that needs to be added to the view root and that replaces/overlays the app UI. In addition to that it returns a Callback object that the app needs to use if it wants to leave full screen mode.
https://developer.android.com/reference/android/webkit/WebChromeClient.html#onShowCustomView(android.view.View,%20android.webkit.WebChromeClient.CustomViewCallback)
* GeckoView will switch itself into fullscreen mode and the app just needs to hide all other UI. Fullscreen mode can be left by calling a method. No callback is needed.

I think we should hide this callback nonsense and everything related inside the `engine-*` component. And just provide a method on the EngineSession to exit full screen mode. Internally the component can do that with the callback or on whatever it needs.

We should add some flag to `browser-session` for the full screen mode. We do not do that in Focus - but it will be needed to correctly restore the full screen state.

### PR DESCRIPTION
**Pull request by jonalmeida**
This was difficult to wrap my head around in the beginning for the known reasons of having different APIs for both the engines. I solved this by having an optional callback for `EngineSession#fullScreenMode` which is called from the SystemWebView.

For GeckoView, we don't need to inform the engine about entering fullscreen mode, so calling `fullScreenMode(true)` does nothing. Only exiting, requires us to notify the engine.

Another difference is that in the `SystemEngineSession` we have access to the current view attached to the EngineSession and we're able to enter and exit Immersive Mode thanks to that, but we're not able to do it for GeckoView, which requires the callers to remember to do it (if they want to?).

### CODE REVIEW

File: components/concept/engine/src/main/java/mozilla/components/concept/engine/EngineSession.kt [hunk 1]
```diff
@@ -179,6 +185,16 @@ abstract class EngineSession(
...
 
+    /**
+     * Enables/disables fullscreen mode. It also invokes a callback when fullscreen mode is exited.
+     * For GeckoView, this shouldn't necessarily be used to
+     *
+     * @param enable Turns fullscreen mode on/off.
+     * @param exitCallback (Optional) Invoked when the engine exits fullscreen mode. This is mostly
+     * needed for only SystemWebView implementations which need to invoke WebView APIs on exiting.
+     */
+    abstract fun fullScreenMode(enable: Boolean, exitCallback: () -> Unit? = {})
```
Comment (pocmo) [inline components/concept/engine/src/main/java/mozilla/components/concept/engine/EngineSession.kt]: If we are able to hide the View inside the engine (see above) then we may be able to do the same for the callback.

File: components/support/ktx/src/main/java/mozilla/components/support/ktx/android/view/View.kt [hunk 2]
```diff
@@ -83,6 +85,69 @@ fun View.hideKeyboard() {
...
+ *
+ * Examples:
+ * ```
+ * toolbar.visibleOrGone(foo != bar)
+ * imageView.visibleOrGone(foo.isEnabled)
+ * ```
+ */
+fun View.visibleOrGone(enabled: Boolean) {
+    this.visibility = if (enabled) View.VISIBLE else View.GONE
+}
```
[the remark this answers is not in the document]
Comment (jonalmeida) [reply components/support/ktx/src/main/java/mozilla/components/support/ktx/android/view/View.kt]: It was mostly for product teams that needed to change multiple View's visibility dependent on some single logic.
Example:
```kotlin
if (enabled) {
  view1.visibility = View.VISIBLE
  view2.visibility = View.VISIBLE
  view3.visibility = View.VISIBLE
} else {
  view1.visibility = View.GONE
  view2.visibility = View.GONE
  view3.visibility = View.GONE
}
```
Instead we could do just this:
```kotlin
view1.visibleOrGone(enabled)
view2.visibleOrGone(enabled)
view3.visibleOrGone(enabled)
```
It's fair though to remove this as well because I can't see many use cases (for example) in focus where we follow that pattern.

File: components/browser/session/src/main/java/mozilla/components/browser/session/Session.kt [hunk 3]
```diff
@@ -47,6 +48,7 @@ class Session(
         fun onTrackerBlocked(session: Session, blocked: String, all: List<String>) = Unit
         fun onLongPress(session: Session, hitResult: HitResult): Boolean = false
         fun onFindResult(session: Session, result: FindResult) = Unit
+        fun onFullScreenChanged(session: Session, enabled: Boolean, view: View? = null) = Unit
```
[the remark this answers is not in the document]
Comment (jonalmeida) [reply components/browser/session/src/main/java/mozilla/components/browser/session/Session.kt]: I was trying to go for giving the client the opportunity to handle the view how ever they wanted, but it makes sense to handle the View placement for them.

In terms of layout, the SystemEngineView is in a FrameLayout, so should we programmatically add a new View above the engine?

File: components/concept/engine/src/main/java/mozilla/components/concept/engine/EngineSession.kt [hunk 4]
```diff
@@ -5,17 +5,22 @@
...
+import android.view.View
 import mozilla.components.support.base.observer.Observable
 import mozilla.components.support.base.observer.ObserverRegistry
 
 /**
  * Class representing a single engine session.
  *
  * In browsers usually a session corresponds to a tab.
+ *
+ * @param fullScreenCallback Holds onto a callback to be invoked when fullscreen mode is exited.
```
Comment (csadilek) [inline components/concept/engine/src/main/java/mozilla/components/concept/engine/EngineSession.kt]: This should be `@property` and docs probably just `A callback to be invoked when...`

File: components/support/ktx/src/main/java/mozilla/components/support/ktx/android/view/View.kt [hunk 5]
```diff
@@ -83,6 +85,69 @@ fun View.hideKeyboard() {
...
+        val window = context.window
+        window.addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
+        window.decorView.systemUiVisibility = (View.SYSTEM_UI_FLAG_LAYOUT_STABLE
+            or View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION
+            or View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN
+            or View.SYSTEM_UI_FLAG_HIDE_NAVIGATION
+            or View.SYSTEM_UI_FLAG_FULLSCREEN
+            or View.SYSTEM_UI_FLAG_IMMERSIVE_STICKY)
+    }
+}
```
[the remark this answers is not in the document]
Comment (jonalmeida) [reply components/support/ktx/src/main/java/mozilla/components/support/ktx/android/view/View.kt]: Yeah, it's a bit weird. Your previous comment about letting the app decide to to call it, means I don't have to put it on a View anymore. I was trying to handle this for them.  ¯\\\_(ツ)\_/¯

File: components/browser/engine-system/src/test/java/mozilla/components/browser/engine/system/SystemEngineSessionTest.kt [hunk 6]
```diff
@@ -393,7 +396,7 @@ class SystemEngineSessionTest {
     }
 
     @Test
-    fun testFindAll() {
+    fun findAll() {
```
[the remark this answers is not in the document]
Comment (jonalmeida) [reply components/browser/engine-system/src/test/java/mozilla/components/browser/engine/system/SystemEngineSessionTest.kt]: Seems fair, I'll stick with the original.

File: components/browser/session/src/main/java/mozilla/components/browser/session/Session.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (pocmo) [reply components/browser/session/src/main/java/mozilla/components/browser/session/Session.kt]: I'd try dropping it into the FrameLayout and hiding the WebView.

File: components/browser/engine-system/src/main/java/mozilla/components/browser/engine/system/SystemEngineSession.kt [hunk 7]
```diff
@@ -211,6 +214,19 @@ class SystemEngineSession(private val defaultSettings: Settings? = null) : Engin
...
+     */
+    override fun fullScreenMode(enable: Boolean, exitCallback: () -> Unit?) {
+        if (enable) {
+            currentView()?.enterToImmersiveMode()
+            fullScreenCallback = exitCallback
+        } else {
+            currentView()?.exitImmersiveModeIfNeeded()
+            fullScreenCallback.invoke()
+        }
+    }
```
[the remark this answers is not in the document]
Comment (jonalmeida) [reply components/browser/engine-system/src/main/java/mozilla/components/browser/engine/system/SystemEngineSession.kt]: Fixed!

**Comment by jonalmeida** [comment]
browser-engine-system: `Could not resolve mozilla:fxa_client_android:0.2.4.`

**Comment by pocmo** [comment]
Meh. Every now and then we have problems with taskcluster or jcenter or plugins.gradle.org not being resolved.

Btw.: You can re-trigger taskcluster by updating the PR title (just add dots..)

**Comment by jonalmeida** [comment]
@pocmo I tried; don't have access. :(

File: components/browser/engine-gecko-beta/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt [hunk 8]
```diff
@@ -180,6 +181,7 @@ class GeckoEngineSession(
     /**
      * See [EngineSession.findNext]
      */
+    @SuppressLint("WrongConstant")
```
Comment (pocmo) [inline components/browser/engine-gecko-beta/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt]: Can you add a comment explaining why this falsely triggers "WrongConstant"?

File: components/support/ktx/src/test/java/mozilla/components/support/ktx/android/view/ViewTest.kt [hunk 9]
```diff
@@ -4,19 +4,19 @@
...
 import org.junit.Assert.assertTrue
 import org.junit.Test
 import org.junit.runner.RunWith
 import org.robolectric.RobolectricTestRunner
 import org.robolectric.RuntimeEnvironment
 import org.robolectric.shadows.ShadowLooper
-import android.util.DisplayMetrics
-import android.view.View
-import org.junit.Assert.assertEquals
-import org.junit.Assert.assertNotEquals
```
Comment (pocmo) [inline components/support/ktx/src/test/java/mozilla/components/support/ktx/android/view/ViewTest.kt]: nit: This file only contains "import" changes now..

File: components/support/ktx/src/main/java/mozilla/components/support/ktx/android/view/View.kt [hunk 10]
```diff
@@ -68,7 +68,7 @@ fun View.showKeyboard(flags: Int = InputMethodManager.SHOW_IMPLICIT) {
  */
 fun View.hideKeyboard() {
     val imm = (context.getSystemService(Context.INPUT_METHOD_SERVICE) ?: return)
-            as InputMethodManager
+        as InputMethodManager
```
Comment (pocmo) [inline components/support/ktx/src/main/java/mozilla/components/support/ktx/android/view/View.kt]: nit: Unrelated

File: components/concept/engine/src/main/java/mozilla/components/concept/engine/EngineSession.kt [hunk 11]
```diff
@@ -179,6 +180,14 @@ abstract class EngineSession(
      */
     abstract fun clearFindMatches()
 
+    /**
+     * Enables/disables fullscreen mode. It also invokes a callback when fullscreen mode is exited.
+     * For GeckoView, this shouldn't necessarily be used to
```
Comment (pocmo) [inline components/concept/engine/src/main/java/mozilla/components/concept/engine/EngineSession.kt]: The comment still mentions the callback :)

File: components/browser/session/src/main/java/mozilla/components/browser/session/Session.kt [hunk 12]
```diff
@@ -232,9 +239,13 @@ class Session(
      * Desktop Mode state, true if the desktop mode is requested, otherwise false.
      */
     var desktopMode: Boolean by Delegates.observable(false) { _, old, new ->
-        notifyObservers(old, new) {
-            onDesktopModeChanged(this@Session, new)
-        }
+        notifyObservers(old, new) { onDesktopModeChanged(this@Session, new) }
+    }
+    /**
+     * The fullscreen state. True if it is in fullscreen mode.
```
Comment (pocmo) [inline components/browser/session/src/main/java/mozilla/components/browser/session/Session.kt]: nit: empty line between `}` and comment of next method.

File: components/browser/engine-system/src/main/java/mozilla/components/browser/engine/system/SystemEngineView.kt [hunk 13]
```diff
@@ -237,6 +249,26 @@ class SystemEngineView @JvmOverloads constructor(
         return false
     }
 
+    internal fun addFullScreenView(view: View, callback: WebChromeClient.CustomViewCallback) {
+        val webView = findViewWithTag<WebView>("webview")
+        webView?.apply { this.visibility = View.GONE }
+
+        fullScreenCallback = callback
+
+        view.tag = "fullscreen"
```
Comment (pocmo) [inline components/browser/engine-system/src/main/java/mozilla/components/browser/engine/system/SystemEngineView.kt]: nit: We should make sure that we do not clash with tags/ids of the consuming apps. For most resources we use a "mozac" prefix and sometimes another "module" prefix (e.g. Strings: `mozac_browser_toolbar_menu_button)`.

File: components/concept/engine/src/main/java/mozilla/components/concept/engine/EngineSession.kt [hunk 14]
```diff
@@ -179,6 +180,14 @@ abstract class EngineSession(
...
+     *
+     * @param enable Turns fullscreen mode on/off.
+     */
+    abstract fun toggleFullScreenMode(enable: Boolean)
+
```
Comment (pocmo) [inline components/concept/engine/src/main/java/mozilla/components/concept/engine/EngineSession.kt]: `toggleFullScreenMode` doesn't seem to be the right thing here: It never goes to fullscreen mode. In GeckoView it only exists full screen mode. In WebView it does nothing. Should this only be a method that exits full screen mode? And shouldn't we invoke the WebView callback in the `browser-engine-system` implementation`

File: components/concept/engine/src/main/java/mozilla/components/concept/engine/EngineSession.kt (hunk 11 above)
Comment (jonalmeida) [reply components/concept/engine/src/main/java/mozilla/components/concept/engine/EngineSession.kt]: Ah, sorry!

File: components/browser/engine-system/src/main/java/mozilla/components/browser/engine/system/SystemEngineView.kt (hunk 13 above)
Comment (jonalmeida) [reply components/browser/engine-system/src/main/java/mozilla/components/browser/engine/system/SystemEngineView.kt]: I used `mosac_system_engine_*`.

File: components/browser/engine-gecko-beta/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt (hunk 8 above)
Comment (jonalmeida) [reply components/browser/engine-gecko-beta/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt]: Sure, it was the linter that complains since there are IntDef values for `@FinderFindFlags` that don't include the default 0 value. I'll add that in a comment by the annotation and I've filed a bug for GV ([bug 1489528][0]).

[0]: https://bugzilla.mozilla.org/show_bug.cgi?id=1489528

File: components/concept/engine/src/main/java/mozilla/components/concept/engine/EngineSession.kt (hunk 14 above)
Comment (jonalmeida) [reply components/concept/engine/src/main/java/mozilla/components/concept/engine/EngineSession.kt]: You're right. Since we're hiding the WebView implementation in the screen this method doesn't follow the original idea. Sorry, I should have thought about this earlier.

File: components/browser/session/src/main/java/mozilla/components/browser/session/Session.kt [hunk 15]
```diff
@@ -122,65 +125,66 @@ class Session(
...
+        notifyObservers(old, new) { onSecurityChanged(this@Session, new) }
     }
 
     /**
      * Configuration data in case this session is used for a Custom Tab.
      */
-    var customTabConfig: CustomTabConfig? by Delegates.observable<CustomTabConfig?>(null) {
-        _, _, new -> notifyObservers { onCustomTabConfigChanged(this@Session, new) }
+    var customTabConfig: CustomTabConfig? by Delegates.observable<CustomTabConfig?>(
+        null) { _, _, new ->
```
[the remark this answers is not in the document]
Comment (jonalmeida) [reply components/browser/session/src/main/java/mozilla/components/browser/session/Session.kt]: Auto-formatting.  ¯\\\_(ツ)\_/¯ 

I'll disable it until I figure out how to stop it from misbehaving.
